packages/@aws-cdk/aws-synthetics/lib/canary.ts

```diff
@@ -0,0 +1,254 @@
+          actions: ['s3:PutObject', 's3:GetBucketLocation'],
+        }),
+        new iam.PolicyStatement({
+          resources: ['*'],
+          actions: ['cloudwatch:PutMetricData'],
+          conditions: {StringEquals: {'cloudwatch:namespace': 'CloudWatchSynthetics'}},
+        }),
+        new iam.PolicyStatement({
+          resources: ['arn:aws:logs:::*'],
+          actions: ['logs:CreateLogStream', 'logs:CreateLogGroup', 'logs:PutLogEvents'],
+        }),
+      ],
+    });
+
+    this.role = props.role ?? new iam.Role(this, 'ServiceRole', {
+      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+      inlinePolicies: {
+        canaryPolicy: policy,
+      },
+    });
+
+    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
+      artifactS3Location: this.artifactBucket.s3UrlForObject(),
+      executionRoleArn: this.role.roleArn,
+      startCanaryAfterCreation: props.startAfterCreation ?? true,
+      runtimeVersion: 'syn-1.0',
+      name: this.physicalName,
+      schedule: {
+        durationInSeconds: props.timeToLive ? String(props.timeToLive.toSeconds()) : '0' ,
+        expression: props.schedule ? props.schedule.expressionString : 'rate(5 minutes)',
+      },
+      failureRetentionPeriod: props.failureRetentionPeriod?.toDays(),
+      successRetentionPeriod: props.successRetentionPeriod?.toDays(),
+      // 🚧 TODO: implement
+      code: {
+        handler: 'index.handler',
+        script: 'exports.handler = async () => {\nconsole.log(\'hello world\');\n};',
+      },
+    });
+
+    this.canaryId = resource.attrId;
+    this.canaryState = resource.attrState;
+    this.canaryName = this.getResourceNameAttribute(resource.ref);
+  }
+
+  /**
+   * Returns a new metric for the canary
+   *
+   * @default avg over 5 minutes
+   */
+  private metric(metricName: string, options?: MetricOptions): Metric {
+    return new Metric({
+      metricName,
+      namespace: 'CloudWatchSynthetics',
+      dimensions: { CanaryName: this.canaryName },
+      statistic: 'avg',
+      ...options,
+    }).attachTo(this);
+  }
+
+  /**
+   * Returns a Duration metric for the canary
+   *
+   * @default avg over 5 minutes
+   */
+  public metricDuration(options?: MetricOptions): Metric {
+    return this.metric('Duration', options);
+  }
+
+  /**
+   * Returns a Success Percent metric for the canary
+   *
+   * @default avg over 5 minutes
+   */
+  public metricSuccessPercent(options?: MetricOptions): Metric {
+    return this.metric('SuccessPercent', options);
+  }
+
+  /**
+   * Returns Failed metric for the canary
+   *
+   * @default avg over 5 minutes
+   */
+  public metricFailed(options?: MetricOptions): Metric {
+    return this.metric('Failed', options);
+  }
+
+  /**
+   * Verifies that the name fits the regex expression: ^[0-9a-z_\-]+$.
+   *
+   * @param name - the given name of the canary
+   */
+  private validateName(name: string) {
+    const regex = new RegExp('^[0-9a-z_\-]+$');
```

Comment:
nit, replace the order of the validation, from a performance perspective, verifying length of a string is an easier action than testing a regex.

Reply:
I've incorporated both comments here. It is valuable to alert the user that their name is too long as a separate error, so I have moved that before the regex. And since `validateName()` is now a `function` in the file scope, I've added the `RegExp` as a `const` variable.